File: libdriveless/tests/src/test_search_frame_process_distance_to_goal.cpp
#include <gtest/gtest.h>
#include "../../include/search_frame.h"
#include "test_utils.h"
#include <cmath>


TEST(TestSearchFrameProcessDistanceToGoal, ProcDistance)
{
    SearchFrame f1(100, 100, {5, 5}, {15, 15});

    const int SIZE = 3 * 100 * 100;

    float *ptr = new float[SIZE];
    memset(ptr, 0x0, sizeof(float) * SIZE);
    f1.copyFrom(ptr);

    f1.processDistanceToGoal(50, -100);

    for (int z = 0; z < 100; z++)
        for (int x = 0; x < 100; x++) {
            float dist = f1.getDistanceToGoal(x, z);
            float dx = x - 50;
            float dz = z + 100;
            float d = sqrt(dx * dx + dz * dz);
            if (std::abs(d - dist) > 0.01)
                FAIL();
        }
            
}

File: libdriveless/tests/python/test_search_frame_process_distance_to_goal.py
import unittest, math, numpy as np
from pydriveless import SearchFrame

class TestSearchFrameDistanceToGoal(unittest.TestCase):
        
    def test_distance_to_goal(self):
        f1 = SearchFrame(100, 100, (-1, -1), (-1, -1))
        costs = np.array([0.0, -1.0], np.float32)
        f1.set_class_costs(costs)
        
        raw_frame = np.full((100, 100, 3), fill_value=0.0, dtype=np.float32)
        f1.set_frame_data(raw_frame)
   
        f1.process_distance_to_goal(50, -100)

        for z in range(100):
            for x in range(100):
                dist = f1.get_distance_to_goal(x, z)
                dx = x - 50
                dz = z + 100
                d = math.sqrt(dx * dx + dz * dz)
                if abs(d - dist) > 0.01:
                    self.fail()

if __name__ == "__main__":
    unittest.main()
